sympy/core/basic.py

```diff
@@ -1077,6 +1077,214 @@ def _eval_subs(self, old, new):
         """
         return None
 
+    def dsubs(self, subsdict, newvars=None, known=None, unknown=None):
+        r"""
+        Function that substitutes for variables in an ordinary differential equation.
+
+        The inputs are -
+
+        ``subsdict``: A dictionary specifying the variable transformations
+        with the old variable as key and the transformation as value.
+
+        ``newvars``: A list of new variables in the transformation rules.
+        It is required if the number of new variables is not the same as
+        the number of old variables.
+
+        ``known``: A list of functions which should be treated as functions
+        whose differentiation rule is known to SymPy.
+
+        ``unknown``: A list of functions which should be treated as functions
+        whose differentiation rule is unknown to SymPy.
+
+        Returns
+        =======
+
+        The transformed differential equation
+
+        Examples
+        ========
+
+        A basic transformation can be achieved by -
+
+        >>> from sympy import symbols, Function, cos
+        >>> f = Function('f')
+        >>> x, a, t = symbols('x a t')
+        >>> eq = f(x).diff(x) + f(x)
+        >>> eq.dsubs({x: t**2})
+        f(t) + Derivative(f(t), t)/(2*t)
```

Comment:
I didn't realise this was a method on Basic. I don't think this should be a Basic method. It should be a function.

Reply:
I think that the user should explicitly specify that they want `f(x)` to become `f(t)`. It's not really a valid transformation since the second `f` is really not the same function as the first one.

Reply:
> I didn't realise this was a method on Basic. I don't think this should be a Basic method. It should be a function.

I'll make it a function in the next commit.

> I think that the user should explicitly specify that they want `f(x)` to become `f(t)`. It's not really a valid transformation since the second `f` is really not the same function as the first one.

Yes, this is the reason why the `known` and `unknown` options exist. By default, all `UndefinedFunction` functions are considered `unknown` which is the reason this happens. This can be changed by adding `f(x)` to the list of known functions.

Reply:
> Yes, this is the reason why the `known` and `unknown` options exist. By default, all `UndefinedFunction` functions are considered `unknown` which is the reason this happens. This can be changed by adding `f(x)` to the list of known functions.

I think this distinction between known and unknown functions is a bit odd and we can come up with something better.

The reason for the distinction is that sometimes the function `f` is considered to be a "function" and sometimes a "variable". In a Physics problem I might have an equation like
```
d T(t)/dt = k
```
where the function `T` represents a physical quantity such as temperature. Then after a transformation like `tau = k t` I would have
```
d T(tau)/dtau = 1
```
Here the function `T` in the second equation is mathematically speaking not the same function. However physically it still represents the temperature so we might want to call it the same name `T`. In this interpretation replacing `T(t)` with `T(tau)` makes sense because we treat `T` is being a variable that has many possible relationships with other variables.

If we instead view `T` as a function then the result of the transformation should be to
```
d T(tau/k)/dtau = 1
```
I think that by default we should prefer this second form of transformation that treats `T` as being strictly an unknown function. However it should also be possible for the user to provide `T(t):T(tau)` to override that.

I think that what I am describing here is essentially equivalent to Maple's distinction between "known" and "unknown" functions.

Reply:
If such transformations are desired, we can include the mapdict that I had originally included. Would that be good enough or is something else required?

Reply:
Its just that the first one is default and the second one can be achieved by adding the function to the list of known functions. If required, we can change this.

Reply:
Can that not be:
```python
In [7]: Derivative(T(tau/k), tau)
Out[7]: 
∂ ⎛ ⎛τ⎞⎞
──⎜T⎜─⎟⎟
∂τ⎝ ⎝k⎠⎠
```
I think that the default should be to do this unless `T(t): T(tau)` is given. Also `known` seems like a very odd way of specifying the distinction to me.

Maybe this should be discussed on the mailing list to see what people think.

Reply:
> Can that not be:
> 
> ```python
> In [7]: Derivative(T(tau/k), tau)
> Out[7]: 
> ∂ ⎛ ⎛τ⎞⎞
> ──⎜T⎜─⎟⎟
> ∂τ⎝ ⎝k⎠⎠
> ```
> 

This would be mathematically wrong right? How can `k` be removed?

> I think that the default should be to do this unless `T(t): T(tau)` is given. Also `known` seems like a very odd way of specifying the distinction to me.
> 
> Maybe this should be discussed on the mailing list to see what people think.

Yeah sure, changing the kwarg name wouldn't be hard. Should I post on the mailing list?

Reply:
> This would be mathematically wrong right? How can `k` be removed?

I guess there should be a factor of `k` in there. What I mean is can it be expressed as a Derivative wrt tau rather than a subs? The main point of `dsubs` is the fact that subs doesn't perform the substitution in the expected way.

> Should I post on the mailing list?

I think so. The question is really what should the signature of `dsubs` be? (You'll need to carefully explain what the differences are and how dchange works and any other examples from other CAS. Don't presume that other people reading the list will understand what you mean.)

Reply:
> I guess there should be a factor of k in there. What I mean is can it be expressed as a Derivative wrt tau rather than a subs? The main point of dsubs is the fact that subs doesn't perform the substitution in the expected way.

In this case, Maple uses the `D` operator. Maybe something of that sort will have to be added.

> I think so. The question is really what should the signature of dsubs be? (You'll need to carefully explain what the differences are and how dchange works and any other examples from other CAS. Don't presume that other people reading the list will understand what you mean.

I'll try to make a list of examples and then post.

Reply:
> In this case, Maple uses the `D` operator. Maybe something of that sort will have to be added.

We definitely want that but adding it is not trivial. The main problem is that functions themselves are not first class objects so e.g. `cos` or `f` is a class rather than an instance. That means that `D(f)` can't be a proper Basic instance with `f` as its arg. Maybe it's possible to work around that though using `Lambda`...

For now though we can still represent the derivative as something like `y*Derivative(f(x/y), x)` rather than as `Subs(Derivative(...), ...)` which I think is nicer.

Reply:
> For now though we can still represent the derivative as something like `y*Derivative(f(x/y), x)` rather than as `Subs(Derivative(...), ...)` which I think is nicer.

What about transformations like `{t: k/tau}` where `tau` is the new variable?

Reply:
What about them? I don't see why that would be any different.

Reply:
> What about them? I don't see why that would be any different.

I'm confused about this. So we replace `Subs(Derivative(T(t), t), t, k/tau)` by `Derivative(T(k/tau), tau) / Derivative(k/tau, tau)`?

Reply:
Actually, the current code can achieve what you are asking for. But there is a small problem due to `doit`. After removing the `doit` from here
https://github.com/sympy/sympy/blob/f74e2eb22db6a69002b1cda2040e82166ec48032/sympy/core/basic.py#L1286
I get the following results -
``` python
In [9]: T = Function('T')
   ...: t, k, tau = symbols('t k tau')
   ...: eq = Eq(T(t).diff(t), k)
   ...: eq
Out[9]: 
d
──(T(t)) = k
dt

In [12]: eq.dsubs({t: tau/k}, newvars=[tau])
Out[12]: 
d
──(T(τ))
dτ
──────── = k
 ∂ ⎛τ⎞
 ──⎜─⎟
 ∂τ⎝k⎠

In [13]: _.doit()
Out[13]: 
  d
k⋅──(T(τ)) = k
  dτ

In [14]: eq.dsubs({t: tau/k}, newvars=[tau], known=[T(t)])
Out[14]:

∂ ⎛ ⎛τ⎞⎞
──⎜T⎜─⎟⎟
∂τ⎝ ⎝k⎠⎠
──────── = k
 ∂ ⎛τ⎞
 ──⎜─⎟
 ∂τ⎝k⎠

In [15]: _.doit()
Out[15]:

⎛ d        ⎞│
⎜───(T(ξ₁))⎟│   τ = k
⎝dξ₁       ⎠│ξ₁=─
                k
```
Avoiding `doit` would not simplify the equation, but using `doit` is converting the `Derivative` of `known` functions to `Subs(Derivative(...), ...)`. So, we must find a way to avoid using `doit` but somehow simplify the equation.

Reply:
I think this should be the default behaviour without needing to specify `known=[T(t)]`. The user should have to provide `T(t): T(tau)` if they want the other kind of substitution.

Maybe the best way to fix the `doit` problem would be to add a `AppliedFunctionDerivative` class that can represent e.g. `f'(x)`.

Reply:
It would just be a Basic class like `AppliedFunctionDerivative(f(t), 0)` where the zero means "differentiate wrt the 0th argument". Then if the 0th argument is a simple symbol like `t` it could evaluate to an ordinary `Derivative`.

Reply:
I just added the `AppliedFuncDerivative` class and tests for it. I was looking in the code for `_derivative_dispatch`, but I'm unable to figure out how to replace the `Subs(Derivative(...), ...)` part by `AppliedFuncDerivative`. Could you suggest anything?

Reply:
The derivative code is complicated. I haven't been through it myself. @Upabjojr ?

Reply:
> The derivative code is complicated.

The derivative code is a mess. It would be nice to clean the code.

Reply:
The problem is that we may want to decide whether it's an ordinary derivative or a matrix/array derivative. So there is some logic to switch between the two. It's really messy and it's probably going to need some cleanup for sure.

Reply:
I think that it's a mistake to mix up array derivatives with ordinary derivatives. There are not really any situations where you would want to mix the two in practice so there should be a separate `ArrayDerivative` class or `array_diff` function or something like that.

Reply:
Since it needs a lot of refactoring, we can use the `AppliedFuncDerivative` only in `dsubs` for now. Or another option can be to let the `Subs(Derivative(..., ...))` part remain and add `AppliedFuncDerivative` in a later PR since `dsubs` is working.

Reply:
Yes, we could do that. We should at least make sure that `dsolve` can recognise an `AppliedDerivative` since that is where we really need `dsubs`.

Reply:
The problem is `dsolve` can't solve ODEs with substitution in them. In fact, for every equation where there is a `Subs` object present, `classify_ode` somehow matches only the lie group method. So, this has to be added first, and then we can make `dsolve` recognize `AppliedFuncDerivative`.
``` python
In [24]: eq = f(x).diff(x)

In [25]: eq
Out[25]: 
d
──(f(x))
dx

In [26]: dsolve(eq)
Out[26]: f(x) = C₁

In [27]: dsolve(eq.subs(x, y**2))
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-27-b156955447c3> in <module>
----> 1 dsolve(eq.subs(x, y**2))

~\sympy\sympy\solvers\ode\ode.py in dsolve(eq, func, hint, simplify, ics, xi, eta, x0, n, **kwargs)
    662             # The key 'hint' stores the hint needed to be solved for.
    663             hint = hints['hint']
--> 664             return _helper_simplify(eq, hint, hints, simplify, ics=ics)
    665
    666 def _helper_simplify(eq, hint, match, simplify=True, ics=None, **kwargs):

~\sympy\sympy\solvers\ode\ode.py in _helper_simplify(eq, hint, match, simplify, ics, **kwargs)
    693             sols = solvefunc.get_general_solution()
    694         else:
--> 695             sols = solvefunc(eq, func, order, match)
    696         if iterable(sols):
    697             rv = [odesimp(eq, s, func, hint) for s in sols]

~\sympy\sympy\solvers\ode\ode.py in ode_lie_group(eq, func, order, match)
   5453     if desols == []:
   5454         raise NotImplementedError("The given ODE " + str(eq) + " cannot be solved by"
-> 5455             + " the lie group method")
   5456     return desols
   5457

NotImplementedError: The given ODE Subs(Derivative(f(x), x), x, y**2) cannot be solved by the lie group method

In [28]: dsolve(Subs(eq, x**2, y).doit())
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-17-1e9d27472b3c> in <module>
----> 1 dsolve(Subs(eq, x**2, y))

~\sympy\sympy\solvers\ode\ode.py in dsolve(eq, func, hint, simplify, ics, xi, eta, x0, n, **kwargs)
    662             # The key 'hint' stores the hint needed to be solved for.
    663             hint = hints['hint']
--> 664             return _helper_simplify(eq, hint, hints, simplify, ics=ics)
    665
    666 def _helper_simplify(eq, hint, match, simplify=True, ics=None, **kwargs):

~\sympy\sympy\solvers\ode\ode.py in _helper_simplify(eq, hint, match, simplify, ics, **kwargs)
    693             sols = solvefunc.get_general_solution()
    694         else:
--> 695             sols = solvefunc(eq, func, order, match)
    696         if iterable(sols):
    697             rv = [odesimp(eq, s, func, hint) for s in sols]

~\sympy\sympy\solvers\ode\ode.py in ode_lie_group(eq, func, order, match)
   5453     if desols == []:
   5454         raise NotImplementedError("The given ODE " + str(eq) + " cannot be solved by"
-> 5455             + " the lie group method")
   5456     return desols
   5457

NotImplementedError: The given ODE Subs(x**2 + Derivative(f(x), x), x**2, y) cannot be solved by the lie group method
```

Reply:
The idea is that `dsubs` computes the substitution without resulting in `Subs` though.

I don't understand what point you are making about adding one thing before another. It isn't necessary to do everything in one PR. This PR is not ready to be merged though. The main issue with it is the fact that it adds a method to Basic but also the code can be simplified or better organised.

Reply:
With the diff
``` python
diff --git a/sympy/core/basic.py b/sympy/core/basic.py
index bb9d54efa3..131edffc46 100644
--- a/sympy/core/basic.py
+++ b/sympy/core/basic.py
@@ -1205,7 +1205,7 @@ def dsubs(self, subsdict, newvars=None, known=None, unknown=None):
         2*sqrt(b)*Derivative(g(b), b) + u**3*Derivative(h(u), (u, 3)) +
         3*u**2*Derivative(h(u), (u, 2)) + u*Derivative(h(u), u) + Derivative(f(x), (x, 2))
         """
-        from sympy.core.function import Function, Derivative, AppliedUndef
+        from sympy.core.function import Subs, Function, Derivative, AppliedUndef
         from sympy.core.symbol import Symbol

         def diffx(e, sym, n):
@@ -1283,7 +1283,13 @@ def diffx(e, sym, n):
                         eq = eq.subs(func.subs(old_var, Function(old_var.name + 'f')(mapdict[old_var])), func.subs(old_var, mapdict[old_var]))
                 eq = eq.subs(Function(var.name + 'f')(mapdict[var]), subsdict[var])

-        return eq.doit()
+        derivatives = eq.atoms(Derivative)
+        for d in derivatives:
+            done = d.doit()
+            if not len(done.atoms(Subs)):
+                eq = eq.subs(d, done)
+
+        return eq

     def xreplace(self, rule):
         """
```
I get the following output.
``` python
In [1]: T = Function('T')
   ...: t, k, tau = symbols('t k tau')
   ...: eq = Eq(T(t).diff(t), k)

In [2]: eq
Out[2]: 
d
──(T(t)) = k
dt

In [3]: eq.dsubs({t: tau/k}, newvars=[tau])
Out[3]: 
  d
k⋅──(T(τ)) = k
  dτ

In [4]: eq.dsubs({t: tau/k}, newvars=[tau], known=[T(t)])
Out[4]: 

  ∂ ⎛ ⎛τ⎞⎞
k⋅──⎜T⎜─⎟⎟ = k
  ∂τ⎝ ⎝k⎠⎠
```
Basically, any applied derivative will not be converted into a `Subs` object. I think this is what we were looking for initially? (https://github.com/sympy/sympy/pull/20979#discussion_r589254926) Is this okay now or should we continue with the `AppliedFuncDerivative` class?

Of course, there are other things that I still have to modify. I will soon add some commits addressing your comments above.

Reply:
> I didn't realise this was a method on Basic. I don't think this should be a Basic method. It should be a function.

Under which module should I make it a function - `solvers.ode` or `core`?